assets/scripts/fe/src/app/components/payroll/payroll-list/payroll-list.component.ts

```diff
@@ -29,4 +33,68 @@ export class PayrollListComponent implements OnInit {
     }
   }
 
-}
+  onToggle(item){
+    // Function trigger upon checkbox toggle
+
+    // Checks if item is already exist in mapCheckedPayroll 
+    if(this.payrollservice.mapCheckedPayroll.has(item.id)){
+      // if item is exist. Remove it from mapped data
+      this.payrollservice.mapCheckedPayroll.delete(item.id);
+    }
+    else{
+      // else, add the item to mapped data
+      this.payrollservice.mapCheckedPayroll.set(item.id, item)
+    } 
+  }
+
+  downloadPDF(){
+    // loops over each item of the mapped data
+    this.payrollservice.mapCheckedPayroll.forEach((value: boolean, key: string) => {
+      // get file name
+      const file_name = this.getFileName(value);
+      // call the download functionality with parameters of item id and file name
+      this.payrollservice.downloadPDF(key, file_name);
+    });
+  }
+
+  getFileName(payrollValue){
+    // Constructing the file name for the pdf
+    const report_phrase = `${payrollValue.date_from} to ${payrollValue.date_to}`;
+    const date_phrase = `${payrollValue.date_from} to ${payrollValue.date_to}`;
+    const employee_name = `${payrollValue.user.first_name} to ${payrollValue.user.last_name}`;
+    const file_name = `payroll of ${employee_name} ${report_phrase}`;
+    return `payroll of ${employee_name} ${date_phrase}`;;
+  }
+
+
+  sendPDF(){
```

Comment:
bad code.

```
private sendToEmail;
  private sending: Boolean = false;

  sendPayroll () {
    // get the list of the selected payroll items
    // then send it to the user's email.
    let payroll = Array.from(this.payrollservice.selectedPayroll.keys());

    if (_.size(payroll) > 0) {
      // enable sending to disable exectuing multiple
      // actions at the same time.
      this.sending = true;

      // send payroll
      this.sendToEmail = this.payrollservice.sendToEmail(payroll)
        .then(resp => { this.sending = false; })
        .catch(err => { this.sending = false; })
      ;
    } 
  }
```

Reply:
by the way, use `lodash` as javascript utility. there's a lot of helpful functions and helpers in that library. it is built in angular already.

`_.size()` is one of them